crates/aptos-protos/proto/aptos/transaction/v1/transaction.proto

```diff
@@ -0,0 +1,415 @@
+  string name = 2;
+  repeated MoveModuleId friends = 3;
+  repeated MoveFunction exposed_functions = 4;
+  repeated MoveStruct structs = 5;
+}
+message MoveFunction {
+  enum Visibility {
+    PRIVATE = 0;
+    PUBLIC = 1;
+    FRIEND = 2;
+  }
+  string name = 1;
+  MoveFunction.Visibility visibility = 2;
+  bool is_entry = 3;
+  repeated MoveFunctionGenericTypeParam generic_type_params = 4;
+  repeated MoveType params = 5;
+  repeated MoveType return = 6;
+}
+
+message MoveStruct {
+  string name = 1;
+  bool is_native = 2;
+  repeated MoveAbility abilities = 3;
+  repeated MoveStructGenericTypeParam generic_type_params = 4;
+  repeated MoveStructField fields = 5;
+}
+
+message MoveStructGenericTypeParam{
+  repeated MoveAbility constraints = 1;
+  bool is_phantom = 2;
+}
+
+message MoveStructField {
+  string name = 1;
+  MoveType type = 2;
+}
+
+message MoveFunctionGenericTypeParam {
+  repeated MoveAbility constraints = 1;
+}
+
+// Next id: 12
+enum MoveTypes {
+  Bool = 0;
+  U8 = 1;
+  U16 = 11;
+  U32 = 12;
+  U64 = 2;
+  U128 = 3;
+  U256 = 13;
+  Address = 4;
+  Signer = 5;
+  Vector = 6; // { items: Box<MoveType> },
+  Struct = 7; //(MoveStructTag),
+  GenericTypeParam = 8; // { index: u16 },
+  Reference = 9; // { mutable: bool, to: Box<MoveType> },
+  Unparsable = 10; //(String),
+}
+
+message MoveType{
+
+  message ReferenceType {
+    bool mutable = 1;
+    MoveType to = 2;
+  }
+
+  MoveTypes type = 1;
+  oneof content {
+    MoveType vector = 3;
+    MoveStructTag struct = 4;
+    uint32 generic_type_param_index = 5;
+    ReferenceType reference = 6;
+    string unparsable = 7;
+  }
+}
+
+enum MoveAbility {
+  COPY = 0;
+  DROP = 1;
+  STORE = 2;
+  KEY = 3;
+}
+
+message WriteSetPayload {
+  WriteSet write_set = 1;
+}
+
+message EntryFunctionId {
+  MoveModuleId module = 1;
+  string name = 2;
+}
+
+message MoveModuleId {
+  string address = 1;
+  string name = 2;
+}
+
+message MoveStructTag {
+  string address = 1;
+  string module = 2;
+  string name = 3;
+  repeated MoveType generic_type_params = 4;
+}
+
+message Signature {
+
+  enum Type {
+    ED25519 = 0;
+    MULTI_ED25519 = 1;
+    MULTI_AGENT = 2;
+  }
+
+  Type type = 1;
+  oneof signature {
+    Ed25519Signature ed25519 = 2;
+    MultiEd25519Signature multi_ed25519 = 3;
+    MultiAgentSignature multi_agent = 4;
+  }
+}
+
+message Ed25519Signature {
+  bytes public_key = 1;
+  bytes signature = 2;
+}
+
+message MultiEd25519Signature {
+  repeated bytes public_keys = 1;
+  repeated bytes signatures = 2;
+  uint32 threshold = 3;
+  repeated uint32 public_key_indices = 4;
+}
+
+message MultiAgentSignature {
+  AccountSignature sender = 1;
+  repeated string secondary_signer_addresses = 2;
+  repeated AccountSignature secondary_signers = 3;
+}
+
+message AccountSignature {
+  enum Type {
+    ED25519 = 0;
+    MULTI_ED25519 = 1;
+  }
+
+  Type type = 1;
+  oneof signature {
+    Ed25519Signature ed25519 = 2;
+    MultiEd25519Signature multi_ed25519 = 3;
+  }
+}
```

Comment:
I'd love for the types generated from this to replace both the API types and the internal types down the line. Especially internal, the JSON version to the API at least _might_ still require some additional types.

Reply:
Well yes, but then the master file should reside close to the Aptos Move VM. This is the source of truth for this kind of definitions.